crates/rspack_core/src/compilation/make/graph_updater/cutout/mod.rs

```diff
@@ -59,23 +60,32 @@ impl Cutout {
           }));
         }
         UpdateParam::ModifiedFiles(files) | UpdateParam::RemovedFiles(files) => {
-          for module in module_graph.modules().values() {
-            // check has dependencies modified
-            if module.depends_on(&files) {
-              // add module id
-              force_build_modules.insert(module.identifier());
-            }
-          }
-          // only failed dependencies need to check
-          for dep_id in &artifact.make_failed_dependencies {
-            let dep = module_graph
-              .dependency_by_id(dep_id)
-              .expect("should have dependency");
-            let info = FactorizeInfo::get_from(dep).expect("should have factorize info");
-            if info.depends_on(&files) {
-              force_build_deps.insert(*dep_id);
-            }
-          }
+          force_build_modules.extend(
+            module_graph
+              .modules()
+              .values()
+              .par_bridge()
+              .filter_map(|module| {
+                if module.depends_on(&files) {
+                  Some(module.identifier())
+                } else {
+                  None
+                }
+              })
+              .collect::<Vec<_>>(),
+          );
+          force_build_deps.extend(
+            module_graph
+              .dependencies()
```

Comment:
The performance degradation may be caused by accessing all dependencies, which can far outnumber the number of modules. Maybe we can generate a new state of FactorizeInfo and save index at module artifact to make it faster.
``` rust
enum FactorizeInfo {
  Success,
  WithMissing { missing: ... }
  Failed { ... }
}

struct MakeArtifact {
   make_failed_dependencies: ...
   has_missing_info_dependencies: ...
}
```

Reply:
Another feasible solution is to use FactorizeInfo.Failed to store missing cases. Treat the case where missing_dependencies is not empty as the case where diagnostics is not empty.

``` rust
// factorize task
fn main() {
  if !factorize_info.diagnostics().is_empty() || !factorize_info.missing_dependencies.is_empty() {
      artifact
        .file_dependencies
        .add_batch_file(&factorize_info.file_dependencies());
      artifact
        .context_dependencies
        .add_batch_file(&factorize_info.context_dependencies());
      artifact
        .missing_dependencies
        .add_batch_file(&factorize_info.missing_dependencies());
      artifact
        .make_failed_dependencies
        .insert(*dependencies[0].id());
  }
}
```

Reply:
@jerrykingxyz Basically, every dependencies have missing dependencies